### PR TITLE
(SERVENG-1097) Add option to target server or workstation

### PR DESCRIPTION
**Pull request by pocketgeek**
All 4 files generated correlate with each other 100%

### CODE REVIEW

**Comment by hsnodgrass** [comment]
Quick question: does using `--target` mean we can only produce mapping files for `server` OR `workstation`? If so, we should work towards a solution that puts both profiles into the same mappings file.

**Comment by pocketgeek** [comment]
> Quick question: does using `--target` mean we can only produce mapping files for `server` OR `workstation`? If so, we should work towards a solution that puts both profiles into the same mappings file.

I was thinking it would be easier to deal with if that stuff was in different files, I can add an option to combine them though.

**Comment by hsnodgrass** [comment]
> > Quick question: does using `--target` mean we can only produce mapping files for `server` OR `workstation`? If so, we should work towards a solution that puts both profiles into the same mappings file.
> 
> I was thinking it would be easier to deal with if that stuff was in different files, I can add an option to combine them though.

I think combining them is a good idea, makes for less files and a less cluttered Hiera hierarchy.

**Comment by pocketgeek** [comment]
> > > Quick question: does using `--target` mean we can only produce mapping files for `server` OR `workstation`? If so, we should work towards a solution that puts both profiles into the same mappings file.
> > 
> > 
> > I was thinking it would be easier to deal with if that stuff was in different files, I can add an option to combine them though.
> 
> I think combining them is a good idea, makes for less files and a less cluttered Hiera hierarchy.

Server and Workstation are added to each of the 4 files, although I don't like the way I did it.  Feels like I should be able to make it more dynamic.

File: lib/abide_dev_utils/xccdf.rb [hunk 1]
```diff
@@ -288,36 +287,46 @@ def diff_controls(other)
...
-                       resolve_control_reference(ctrl).xpath('./xccdf:title').text
-                     else
-                       ctrl.send(idx_sym.to_sym)
-                     end
-              ary << "#{item}"
+
+        objserver = {}
+        objworkstation = {}
+
+        ['workstation', 'server'].each do |t|
```
Comment (hsnodgrass) [inline lib/abide_dev_utils/xccdf.rb:294]: Instead of hardcoding this, when you loop through `profiles` it yields a `Profile` object. You can get the profile name from the object itself. Take a look at https://github.com/puppetlabs/abide_dev_utils/blob/eaf247b638825afe32428301ca9ab8fad1cac060/lib/abide_dev_utils/xccdf.rb#L635

File: lib/abide_dev_utils/cli/xccdf.rb [hunk 2]
```diff
@@ -43,10 +43,11 @@ def execute(xccdf_file)
         end
         @data[:console] = true if @data[:dir].nil?
         @data[:type] = 'cis' if @data[:type].nil?
+        @data[:parent_key_prefix] = @data[:prefix]
         @data[:parent_key_prefix] = '' if @data[:parent_key_prefix].nil?
         hfile = AbideDevUtils::XCCDF.gen_map(xccdf_file, **@data)
         mapping_dir = File.dirname(hfile.keys[0]) unless @data[:dir].nil?
-        AbideDevUtils::Output.simple("Creating directory #{mapping_dir}") unless @data[:quiet] || @data[:console] || @data[:dir].nil?
+        AbideDevUtils::Output.simple("Creating directory #{mapping_dir}") unless @data[:quiet] || @data[:console] || @data[:dir].nil? || File.directory?(mapping_dir)
```
Comment (hsnodgrass) [inline lib/abide_dev_utils/cli/xccdf.rb:50]: Adding the `File.directory?` conditional here doesn't actually stop the dir creation attempt, it just stops the output.

File: lib/abide_dev_utils/cli/xccdf.rb [hunk 3]
```diff
@@ -43,10 +43,11 @@ def execute(xccdf_file)
         end
         @data[:console] = true if @data[:dir].nil?
         @data[:type] = 'cis' if @data[:type].nil?
+        @data[:parent_key_prefix] = @data[:prefix]
```
Comment (hsnodgrass) [inline lib/abide_dev_utils/cli/xccdf.rb:46]: Why not just keep the assignment to `@data[:parent_key_prefix]` in the first place?

File: lib/abide_dev_utils/cli/xccdf.rb [hunk 4]
```diff
@@ -32,7 +32,7 @@ def initialize
         options.on('-d [DIR]', '--files-output-directory [DIR]', 'Directory to save files data/mappings by default') { |d| @data[:dir] = d }
         options.on('-q', '--quiet', 'Show no output in the terminal') { @data[:quiet] = true }
         options.on('-p [PREFIX]', '--parent-key-prefix [PREFIX]', 'A prefix to append to the parent key') do |p|
-          @data[:parent_key_prefix] = p
+          @data[:prefix] = p
```
Comment (hsnodgrass) [inline lib/abide_dev_utils/cli/xccdf.rb:35]: This just gets reassigned to `@data[:parent_key_prefix]` on line 46, why the change?

File: lib/abide_dev_utils/xccdf.rb [hunk 5]
```diff
@@ -288,36 +287,46 @@ def diff_controls(other)
...
-              ary << "#{item}"
+
+        objserver = {}
+        objworkstation = {}
+        #mappings = [framework, index]
+        #mappings.unshift(key_prefix) unless key_prefix.empty?
+        #output = { mappings.join('::') => {} }
+        output = {}
+
+          c_map = profiles.each_with_object({}) do |profile, obj|
```
Comment (hsnodgrass) [inline lib/abide_dev_utils/xccdf.rb:298]: Indentation is off here

**Review by hsnodgrass — CHANGES_REQUESTED** [review]
Just a few comments

File: lib/abide_dev_utils/cli/xccdf.rb (hunk 2 above)
Comment (pocketgeek) [reply lib/abide_dev_utils/cli/xccdf.rb:50]: Which is what I wanted.

File: lib/abide_dev_utils/xccdf.rb (hunk 5 above)
Comment (pocketgeek) [reply lib/abide_dev_utils/xccdf.rb]: Fixed

File: lib/abide_dev_utils/xccdf.rb [hunk 6]
```diff
@@ -236,7 +236,6 @@ def gen_map(dir: nil, type: 'CIS', parent_key_prefix: '', **_)
         else
           mapping_dir = ''
         end
-        parent_key_prefix = parent_key_prefix.nil? ? nil : ''
```
Comment (hsnodgrass) [inline lib/abide_dev_utils/xccdf.rb:239]: This was in there because if you explicitly pass a value of `nil` into that parameter, it will cause exceptions to be thrown later in the code.

**Review by hsnodgrass — CHANGES_REQUESTED** [review]

File: lib/abide_dev_utils/cli/xccdf.rb (hunk 4 above)
Comment (pocketgeek) [reply lib/abide_dev_utils/cli/xccdf.rb:35]: Was keeping with the trend of the stuff in [] matching @data[:]

File: lib/abide_dev_utils/xccdf.rb (hunk 6 above)
Comment (pocketgeek) [reply lib/abide_dev_utils/xccdf.rb:239]: Fair enough

File: lib/abide_dev_utils/cli/xccdf.rb (hunk 3 above)
Comment (pocketgeek) [reply lib/abide_dev_utils/cli/xccdf.rb:46]: Yeah, I guess that was kinda silly.

**Review by hsnodgrass — CHANGES_REQUESTED** [review]
That `nil` check should be put back

**Comment by pocketgeek** [comment]
I did, except I didn't save the file....  

Fixed.

**Review by hsnodgrass — APPROVED** [review]
LGTM
